### ISSUE TITLE
Feat Prerequisite Standardization

### ISSUE THREAD

**Issue by jerenda** [openlegend/core-rules#70]
I'm working on an edit to alphabetize all of the Attributes listed in the Banes and in the Feats sections. The prerequisites for the feat Battlefield Defender are styled "Agility 4 OR Might 4 OR Protection 4", but the prerequisites for the feat Ageless are styled "Alteration, Creation, Entropy, or Protection 9". (Note that the actual text is not alphabetized.) Which one would you prefer? I can standardize it while I'm doing this.

**Comment by raihje** [comment]
It'd probably be best to alphabetize everything (order of feats, order of prerequisites) Some feats say Agility 1 / Might 1 some say it with a capital OR some say it like you stated above so something should be done to standardize that as well. Just my two cents.

**Comment by jerenda** [comment]
I assumed that the feats were already alphabetized, but I can double-check that if you noticed something off.

Now that I'm reading through the rest, the OR seems to be the standard. It's just Ageless that's off, which I've taken care of.

But a lot of other things about the feats seem really non-standard. The tags are screwy, with a Might and Agility tag but not an Entropy tag, for instance, and they're not applied systematically. Given that tags seem to be primarily categorizational information for the programmer, that seems a bit counter-productive. And they have some other weird details, like the fact that the prereqs for Craft Beneficial or Baleful Item (I - IV) are "Any Supernatural Attribute 5 / Learning 5 / Creation 5" when Creation is a Supernatural attribute already, or that Energy Resistance prereqs read "Energy OR 4/5/7??" I think modifying all of those details is a little outside the scope of my current alphabetize-only edit, but I might have to return to study the feats in more detail later.

**Comment by brianfeister** [comment]
You should actually put this on pause altogether because I need to do something fancier and more complex with the YAML file to work on improving the pre-requisites for a future character-builder

If you are comfortable with enforcing proper code style and indentation, want to volunteer to do lots of work, you can follow this style:

https://gist.github.com/brianfeister/3d5ccc202b566ca06ece00b984662591

**Comment by brianfeister** [comment]
TLDR; There's more than just a small fix that's needed here so we can either leave it alone or do the whole thing.

**Comment by jerenda** [comment]
Hah, ok. I've already finished the edit I intended to do tonight, so I'll just propose the two changes (Banes/Feats) separately and you can decide if you want to accept them or not. I'm the kind of person who takes tiny details very seriously, and I'm considering running my personal world in this system, so it's entirely likely that I'll end up doing the "lots of work" option. XD

**Comment by raihje** [comment]
I was going to comment that I'd do feats, while he did the other ones but it just sat here untyped and now there's more comments.

I'm fine doing this over the weekend, I have a halloween party Saturday night, but other than that I'm unemployed at the moment and sick of having nothing but resumes to hand out all day.

**Comment by jerenda** [comment]
Hey, my weekends are mad. I probably won't take another look at this until Monday. Go for it.

**Comment by raihje** [comment]
I'll do the style for the feats, hopefully he'll have an example for banes/boons if there's anything different before I dive in making them identical to feats.

**Comment by raihje** [comment]
@brianfeister I've updated my fork with the Energy Resistance update and the style guideline for feats.yml Let me know what you think.

I only added tiers for feats that have requirements that change for each tier. If a feat had the same requirement or 'None; as a requirement I didn't add tiers. For 'None' I left it as is.
[raihje / core-rules/feats/feats.yml](https://github.com/raihje/core-rules/blob/master/feats/feats.yml)

**Comment by brianfeister** [comment]
FYI for everyone, I'm technically on vacation and trying to avoid being slain for violation of the rules of not working 24/7 on vacation.

@raihje you're right, we need an `OR`, the problem is I really don't want to move away from YAML and I'm trying to think of the best syntax to support `OR` in YAML.

Let's do it this way, everyone. Each child property of `tier{x}` is required, if that property is itself an array, then the array represents `OR` and any of the prerequisites will yield valid / prerequisite met.

So in this example, `Some Other Feat Name` -- AND -- (`Alteration: 5` OR `Creation: 5`) are required

```
- !
  name: Some Feat Name
  prerequisite:
  - tier1:
    - Attribute: 
      - Alteration: 5
      - Creation: 5
    - Feat:
      - Some Other Feat Name
```

But in this example all 3 of the following are required: `Some Other Feat Name` -- AND -- `Alteration: 5` -- AND -- `Creation: 5` are required

```
- !
  name: Some Feat Name
  prerequisite:
  - tier1:
    - Attribute: 
      - Alteration: 5
    - Attribute: 
      - Creation: 5
    - Feat:
      - Some Other Feat Name
```

**Comment by brianfeister** [comment]
Let me know if that works for you folks and if so, who wants to tackle it. My availability will continue to be limited so I can only drop in for short stints here and there.

**Comment by raihje** [comment]
The only thing we're missing here is OR FEAT Such iast learning 3 OR knowledge I for crafting.

**Comment by brianfeister** [comment]
that would be handled the same way, so 

```
- !
  name: Some Feat Name
  prerequisite:
  - tier1:
    - Feat:
      - First Optional Feat Name
      - Second Optional Feat Name
```

**Comment by raihje** [comment]
Ah you're right sorry. I just woke up lol. However that's for "feat or feat" how does that solve the issue of some things like Crafting which can use either an attribute (learning 3) or a feat (Knowledge 1) this is great if a feat has optional feat requirements.

**Comment by brianfeister** [comment]
Actually, you were right... I'm thinking about it. Are there any other examples of FEAT OR ATTRIBUTE?

**Comment by raihje** [comment]
I think Baneful Item is "Any supernatural" or "Learning" so that's another case to possible consider (I was just going to do the following)
`- !
  name: Some Feat Name
  prerequisite:
  - tier1:
    - Attribute: 
      - Any Supernatural: 5
      - Learning: 5
`
As for feats that have attribute or feat requirements, I'm not sure I'll check and edit this comment.

**Comment by brianfeister** [comment]
I suppose since the game only has TWO things that can be prerequisites (feats and attributes, there's nothing else), we could do:

```
- !
  name: Some Feat Name
  prerequisite:
  - tier1:
    - FeatAttribute:
      - Feat: First Optional Feat Name
      - Attribute: 
        - Creation: 5
```

**Comment by raihje** [comment]
That's alright then. If there's no feats just stick with - Attribute over the FeatAttribute? Likewise for mandatory feats put it in it's own - Feat.

Crafting afaik with a quick skim is the only feat like that anyway. The rest are "Ability to invoke X" or "None" or some mixture of attributes.
For ability to invoke boon/bane I wasn't putting it inside any sort of - Attribute or - Feat it was just on it's own under - tierX is that alright?

**Comment by brianfeister** [comment]
I guess "Ability to Invoke" should probably get lumped under `Other` as a thing we don't really filter on within application logic:

```
- !
  name: Some Feat Name
  prerequisite:
  - tier1:
    - Other:
      - Ability to Invoke ___
    - FeatAttribute:
      - Feat: First Optional Feat Name
      - Attribute: 
        - Creation: 5
```

**Comment by raihje** [comment]
Alright, that's fine. - Other for "None" or "Ability to Invoke Bane/Boon"
Mandatory Feats in their own - Feat. Mandatory Attributes in their own - Attribute choices between attributes are an array within - Attribute and FeatAttribute for Crafting since it's Attribute OR Feat.

Does the above sound correct? If so, I'll implement it immediately and issue a PR.

**Comment by brianfeister** [comment]
Sounds correct to me, thanks @raihje !

**Comment by raihje** [comment]
One last question. Is it okay to put Any: X and Any <Category>: X
for example for feats with Any 1 / Any 2 etc..
`- Attribute:
    - Any: 1
- Attribute:
   - Any Supernatural: 1`

**Comment by brianfeister** [comment]
Yeah I think that's fine, we will need to treat those as separate "attributes" but I think that's the right way to handle them

**Comment by raihje** [comment]
Excellent, I'll have this done in the next half hour then I need to shop-vac my rug since my hot water tank is leaking lol.

**Comment by brianfeister** [comment]
`Any` `Any Physical` `Any Supernatural` will each be their own thing and we would probably just include them behind the scenes when you have a Supernatural attribute active for filtering

**Comment by brianfeister** [comment]
LOL, sounds good

**Comment by raihje** [comment]
@brianfeister Alright I made PR #86 if you like it feel free to close this issue? Pretty sure the alphabetizing commits have already been merged from another PR.

**Comment by brianfeister** [comment]
Closing re: #86
